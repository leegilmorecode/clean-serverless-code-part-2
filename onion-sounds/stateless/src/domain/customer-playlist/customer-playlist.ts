import * as customerPlaylistCreatedEvent from '@events/customer-playlist-created';
import * as songAddedToPlaylistEvent from '@events/song-added-to-playlist';

import {
  CreateCustomerPlaylistProps,
  NewCustomerPlaylistProps,
} from '@models/customer-playlist-types';

import { CustomerPlaylistDto } from '@dto/customer-playlist';
import { Entity } from '@entity/entity';
import { MaxPlaylistSizeError } from '@errors/max-playlist-size-error';
import { schema } from '@schemas/customer-playlist.schema';

/**
 * Customer Playlist Domain Entity
 *
 * A customer playlist is an entity which allows customers to manage one or more playlists.
 */
export class CustomerPlaylist extends Entity<CreateCustomerPlaylistProps> {
  private constructor({
    id,
    created,
    updated,
    ...props
  }: CreateCustomerPlaylistProps) {
    super(props, id, created, updated);
  }

  // create an instance of a new customer playlist through a factory method
  public static createPlaylist(
    props: NewCustomerPlaylistProps
  ): CustomerPlaylist {
    const customerAccountProps: CreateCustomerPlaylistProps = {
      playlistName: props.playlistName,
      songIds: [],
    };

    const instance: CustomerPlaylist = new CustomerPlaylist(
      customerAccountProps
    );

    instance.validate(schema);

    instance.addDomainEvent({
      event: instance.toDto(),
      eventName: customerPlaylistCreatedEvent.eventName,
      source: customerPlaylistCreatedEvent.eventSource,
      eventSchema: customerPlaylistCreatedEvent.eventSchema,
      eventVersion: customerPlaylistCreatedEvent.eventVersion,
    });

    return instance;
  }

  // add a song to the playlist
  public addSongToPlaylist(songId: string): void {
    if (this.props.songIds.length >= 4) {
      throw new MaxPlaylistSizeError('the maximum playlist length is 4');
    }

    this.props.songIds.push(songId);

    this.setUpdatedDate();
    this.validate(schema);

    this.addDomainEvent({
      event: this.toDto(),
      eventName: songAddedToPlaylistEvent.eventName,
      source: songAddedToPlaylistEvent.eventSource,
      eventSchema: songAddedToPlaylistEvent.eventSchema,
      eventVersion: songAddedToPlaylistEvent.eventVersion,
    });
  }

  // create a dto based on the domain instance
  public toDto(): CustomerPlaylistDto {
    return {
      id: this.id,
      created: this.created,
      updated: this.updated,
      playlistName: this.props.playlistName,
      songIds: this.props.songIds,
    };
  }

  // create a domain object based on the dto
  public static toDomain(raw: CustomerPlaylistDto): CustomerPlaylist {
    const instance = new CustomerPlaylist(raw);
    return instance;
  }
}
