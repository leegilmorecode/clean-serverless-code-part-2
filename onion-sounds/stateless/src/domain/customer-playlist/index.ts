export { CustomerPlaylist } from './customer-playlist';
