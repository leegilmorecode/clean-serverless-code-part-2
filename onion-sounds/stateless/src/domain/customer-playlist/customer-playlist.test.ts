import {
  CustomerPlaylistProps,
  NewCustomerPlaylistProps,
} from '@models/customer-playlist-types';

import { CustomerPlaylist } from '@domain/customer-playlist';
import { CustomerPlaylistDto } from '@dto/customer-playlist';

let customerPlaylist: CustomerPlaylistDto = {
  created: '2022-01-01T00:00:00.000Z',
  id: 'f39e49ad-8f88-448f-8a15-41d560ad6d70',
  updated: '2022-01-01T00:00:00.000Z',
  songIds: ['123', '345'],
  playlistName: 'testplaylist',
};

describe('customer-playlist', () => {
  beforeAll(() => {
    jest.useFakeTimers().setSystemTime(new Date('2022-01-01'));
  });

  afterAll(() => {
    jest.clearAllTimers();
    jest.clearAllMocks();
  });

  describe('create playlist', () => {
    it('should fail creating a new account with an invalid customerPlaylist', () => {
      // arrange
      const newCustomerPlaylist: NewCustomerPlaylistProps = {
        ...customerPlaylist,
        playlistName: 'Â±',
      };

      // act / assert
      expect(() =>
        CustomerPlaylist.createPlaylist(newCustomerPlaylist)
      ).toThrowErrorMatchingInlineSnapshot(
        `"[{\\"instancePath\\":\\"/playlistName\\",\\"schemaPath\\":\\"#/properties/playlistName/pattern\\",\\"keyword\\":\\"pattern\\",\\"params\\":{\\"pattern\\":\\"^[a-zA-Z]+$\\"},\\"message\\":\\"must match pattern \\\\\\"^[a-zA-Z]+$\\\\\\"\\"}]"`
      );
    });

    it('should create the new account successfully with a valid customerPlaylist', () => {
      // arrange
      const newCustomerPaylist: NewCustomerPlaylistProps = {
        ...customerPlaylist,
        playlistName: 'testplaylist',
      };

      // act
      const newPlaylist = CustomerPlaylist.createPlaylist(newCustomerPaylist);

      // assert
      expect(newPlaylist).toMatchSnapshot();
    });
  });

  describe('toDto', () => {
    it('should create the correct dto', () => {
      // arrange
      const newCustomerPlaylist: NewCustomerPlaylistProps = {
        playlistName: 'testplaylist',
      };

      // act
      const newPlaylist = CustomerPlaylist.createPlaylist(newCustomerPlaylist);

      // assert
      expect(newPlaylist.toDto()).toMatchSnapshot();
    });
  });

  describe('toDomain', () => {
    it('should create a domain object based on a dto', () => {
      // arrange
      const customerPlaylistProps: CustomerPlaylistProps = {
        ...customerPlaylist,
      };

      // act
      const playlist: CustomerPlaylist = CustomerPlaylist.toDomain(
        customerPlaylistProps
      );

      // assert
      expect(playlist).toMatchSnapshot();
    });
  });

  describe('addSongToPlaylist', () => {
    it('should successfully add the song to the playlist', () => {
      // arrange
      const newCustomerPaylist: NewCustomerPlaylistProps = {
        ...customerPlaylist,
        playlistName: 'testplaylist',
      };
      const newPlaylist = CustomerPlaylist.createPlaylist(newCustomerPaylist);

      // act
      newPlaylist.addSongToPlaylist('songone');

      // assert
      expect(newPlaylist).toMatchInlineSnapshot(`
CustomerPlaylist {
  "_created": "2022-01-01T00:00:00.000Z",
  "_domainEvents": Array [
    Object {
      "event": Object {
        "created": "2022-01-01T00:00:00.000Z",
        "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
        "playlistName": "testplaylist",
        "songIds": Array [
          "songone",
        ],
        "updated": "2022-01-01T00:00:00.000Z",
      },
      "eventDateTime": "2022-01-01T00:00:00.000Z",
      "eventName": "CustomerPlaylistCreated",
      "eventVersion": "1",
      "source": "com.customer-account-onion",
    },
    Object {
      "event": Object {
        "created": "2022-01-01T00:00:00.000Z",
        "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
        "playlistName": "testplaylist",
        "songIds": Array [
          "songone",
        ],
        "updated": "2022-01-01T00:00:00.000Z",
      },
      "eventDateTime": "2022-01-01T00:00:00.000Z",
      "eventName": "SongAddedToPlaylist",
      "eventVersion": "1",
      "source": "com.customer-account-onion",
    },
  ],
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "playlistName": "testplaylist",
    "songIds": Array [
      "songone",
    ],
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`);
    });

    it('should throw an error if there are more than 4 songs on the playlist', () => {
      // arrange
      const newCustomerPaylist: NewCustomerPlaylistProps = {
        ...customerPlaylist,
        playlistName: 'testplaylist',
      };
      const newPlaylist = CustomerPlaylist.createPlaylist(newCustomerPaylist);

      // act
      newPlaylist.addSongToPlaylist('songone');
      newPlaylist.addSongToPlaylist('songtwo');
      newPlaylist.addSongToPlaylist('songthree');
      newPlaylist.addSongToPlaylist('songfour');

      // assert
      expect(() =>
        newPlaylist.addSongToPlaylist('songfive')
      ).toThrowErrorMatchingInlineSnapshot(
        `"the maximum playlist length is 4"`
      );
    });
  });
});
