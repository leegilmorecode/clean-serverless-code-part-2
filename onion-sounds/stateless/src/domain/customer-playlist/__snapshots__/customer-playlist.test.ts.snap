// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customer-playlist create playlist should create the new account successfully with a valid customerPlaylist 1`] = `
CustomerPlaylist {
  "_created": "2022-01-01T00:00:00.000Z",
  "_domainEvents": Array [
    Object {
      "event": Object {
        "created": "2022-01-01T00:00:00.000Z",
        "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
        "playlistName": "testplaylist",
        "songIds": Array [],
        "updated": "2022-01-01T00:00:00.000Z",
      },
      "eventDateTime": "2022-01-01T00:00:00.000Z",
      "eventName": "CustomerPlaylistCreated",
      "eventVersion": "1",
      "source": "com.customer-account-onion",
    },
  ],
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "playlistName": "testplaylist",
    "songIds": Array [],
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;

exports[`customer-playlist toDomain should create a domain object based on a dto 1`] = `
CustomerPlaylist {
  "_created": "2022-01-01T00:00:00.000Z",
  "_domainEvents": Array [],
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "playlistName": "testplaylist",
    "songIds": Array [
      "123",
      "345",
    ],
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;

exports[`customer-playlist toDto should create the correct dto 1`] = `
Object {
  "created": "2022-01-01T00:00:00.000Z",
  "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "playlistName": "testplaylist",
  "songIds": Array [],
  "updated": "2022-01-01T00:00:00.000Z",
}
`;
